onload/cplane/main.go

```diff
@@ -0,0 +1,133 @@
+// SPDX-License-Identifier: MIT
+// SPDX-FileCopyrightText: (c) Copyright 2023 Advanced Micro Devices, Inc.
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+const (
+	cplaneServerPathFilePath       = "/sys/module/onload/parameters/cplane_server_path"
+	cplaneServerParamsFilePath     = "/sys/module/onload/parameters/cplane_server_params"
+	critclFilePath                 = "/usr/bin/crictl"
+	crioPrefix                     = "cri-o://"
+	cplaneServerParamsFormatString = "exec %s /opt/onload/sbin/onload_cp_server -K"
+
+	timeoutTime = 5 * time.Second
+)
+
+var (
+	podName      string
+	podNamespace string
+
+	containerID string
+)
+
+func main() {
+
+	podName = os.Getenv("POD_NAME")
+	podNamespace = os.Getenv("POD_NAMESPACE")
+
+	cfg, err := config.GetConfig()
+	if err != nil {
+		glog.Fatalf("Failed to get client config (%v)", err)
+	}
+
+	c, err := client.New(cfg, client.Options{})
+	if err != nil {
+		glog.Fatalf("Failed to create client (%v)", err)
+	}
+	pod := &corev1.Pod{}
+
+	// The information about the pod may be out of date when it is received,
+	// so we enter a "while" loop to keep checking about the pod and container
+	// until they report as running.
+	for {
+		err = c.Get(context.Background(), client.ObjectKey{
+			Namespace: podNamespace,
+			Name:      podName,
+		}, pod)
+		if err != nil {
+			glog.Fatalf("Failed to find pod %q in namespace %q (%v)",
+				podName, podNamespace, err)
+		}
+
+		// We only expect a single container within the pod, so error out if a
+		// different number is found
+		if len(pod.Status.ContainerStatuses) != 1 {
+			glog.Fatalf("Expecting 1 containerStatus in pod %q, found %d",
+				podName, len(pod.Status.ContainerStatuses))
+		}
+		containerStatus := pod.Status.ContainerStatuses[0]
+
+		if !containerStatus.Ready {
+			glog.Warningf("Container is not ready. Will retry in %v",
+				timeoutTime)
+			time.Sleep(timeoutTime)
+			continue
+		}
+
+		if !*containerStatus.Started {
+			glog.Warningf("Container has not started. Will retry in %v",
+				timeoutTime)
+			time.Sleep(timeoutTime)
+			continue
+		}
+
+		if containerStatus.State.Running == nil {
+			glog.Warningf("Container is not running. Will retry in %v",
+				timeoutTime)
+			time.Sleep(timeoutTime)
+			continue
+		}
```

Comment:
Go supports nested functions, so this could be DRY by just passing the state string to a err handler `if ... warnAndWait('ready')`